### PR TITLE
Implement shard shrinking

### PR DESCRIPTION
**Pull request by smakalias**
## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 1]
```diff
@@ -236,6 +271,138 @@ private Session session() {
...
+        if (isPartitioned) {
+            Optional<PartitionName> partitionName = analysis.partitionName();
+            assert partitionName.isPresent() : "Resizing operations for partitioned tables " +
+                                               "are only supported at partition level";
+            sourceIndexName = partitionName.get().asIndexName();
+        } else {
+            sourceIndexName = table.ident().indexName();
+        }
+        String targetIndexName = sourceIndexName + "_shrinked_" + System.currentTimeMillis();
+        //String targetIndexName = ".resized" + sourceIndexName + String.valueOf(System.currentTimeMillis());
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: we should use a name which cannot be used by a user defined table. a `.resized.` prefix would work, a `_shrinked_` suffix not.

File: sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java [hunk 2]
```diff
@@ -22,219 +22,31 @@
...
-public class RenameTableClusterStateExecutor extends DDLClusterStateTaskExecutor<RenameTableRequest> {
-
-    private static final IndicesOptions STRICT_INDICES_OPTIONS = IndicesOptions.fromOptions(false, false, false, false);
-
-    private final Logger logger;
-    private final IndexNameExpressionResolver indexNameExpressionResolver;
-    private final MetaDataIndexAliasesService metaDataIndexAliasesService;
-    private final NamedXContentRegistry xContentRegistry;
-    private final DDLClusterStateService ddlClusterStateService;
+public class RenameTableClusterStateExecutor extends AbstractRenameTableClusterStateExecutor {
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java]: I'd remove the Abstract one and just extend the Exchange class from this one. Currently this child inherit everything from the abstract...

File: sql/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java [hunk 3]
```diff
@@ -687,9 +687,60 @@ public void testAlterShardsOfPartitionedTable() throws Exception {
...
+    }
+
+    @Test
+    public void testAlterShardsTable() throws Exception {
+        execute("create table quotes (id integer, quote string, date timestamp) " +
+                "clustered into 3 shards with (number_of_replicas='0-all', \"blocks.write\"=true)");
+
+        execute("alter table quotes set (number_of_shards=1)");
+        GetSettingsResponse settingsResponse = client().admin().indices().prepareGetSettings("quotes").execute().get();
+        assertThat(settingsResponse.getSetting("quotes", IndexMetaData.SETTING_NUMBER_OF_SHARDS), is("1"));
```
Comment (seut) [inline sql/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java]: why are you using the es api here? the sql below is checking the same...

File: sql/src/main/java/io/crate/metadata/cluster/ExchangeTableNameClusterStateExecutor.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+        final boolean isPartitioned = request.isPartitioned();
+        final boolean validateClosed = false;
+        RelationName xRelationName = request.sourceTableIdent();
+        RelationName yRelationName = request.targetTableIdent();
+        RelationName backupRelationName = RelationName.fromIndexName(yRelationName.indexName() + "_backup");
+        ClusterState state = currentState;
+
+        // y -> backup
+        state = doExecute(state, yRelationName, backupRelationName, isPartitioned, validateClosed);
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/cluster/ExchangeTableNameClusterStateExecutor.java]: imho we must operate on index names here, the operation should not have to know about partitions...

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 1 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: done

File: sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java (hunk 2 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java]: not relevant anymore

File: sql/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java (hunk 3 above)
Comment (smakalias) [reply sql/src/test/java/io/crate/integrationtests/DDLIntegrationTest.java]: removed es api

File: sql/src/main/java/io/crate/metadata/cluster/ExchangeTableNameClusterStateExecutor.java (hunk 4 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/metadata/cluster/ExchangeTableNameClusterStateExecutor.java]: done

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 5]
```diff
@@ -236,6 +271,138 @@ private Session session() {
...
+        final int currentNumberOfShards = clusterService.state().metaData().index(indexName).getNumberOfShards();
+        if (currentNumberOfShards == targetNumberOfShards) {
+            throw new IllegalArgumentException(
+                String.format("Table/partition is already allocated <%d> shards", currentNumberOfShards));
+        } else if (targetNumberOfShards > currentNumberOfShards){
+            throw new IllegalArgumentException("Increasing the number of shards is not supported");
+        }
+
+        final Boolean blocks_write = clusterService.state().metaData().index(indexName)
+            .getSettings().getAsBoolean(INDEX_BLOCKS_WRITE_SETTING.getKey(), false);
```
[the remark this answers is not in the document]
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: let's discuss this as in the ES documentation it is mentioned that in order to prepare index for shrinking you need to mark it as read only: `"index.blocks.write": true`

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 6]
```diff
@@ -236,6 +285,180 @@ private Session session() {
...
+        assert !clusterService.state().nodes().getDataNodes().isEmpty() : "No Data nodes available " +
+                                                                          "for index resize operation";
+        return clusterService.state().nodes().getDataNodes().valuesIt().next();
+    }
+
+    private CompletableFuture<Long> restrictAllocationToNode(String nodeName, String... indices) {
+        Settings settings = Settings.builder()
+            .put(INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name", nodeName)
+            // added temporarily for convenience - todo: remove
+            //.put(SETTING_BLOCKS_WRITE, "true")
```
Comment (smakalias) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: will be removed

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 7]
```diff
@@ -236,6 +285,180 @@ private Session session() {
...
+            .build();
+
+        return updateSettings(settings, indices);
+    }
+
+    private CompletableFuture<Long> revertAllocation(String... indices) {
+        Settings settings = Settings.builder()
+            .putNull(INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name")
+            // added temporarily for convenience - todo: remove
+            //.put(SETTING_BLOCKS_WRITE, "false")
```
Comment (smakalias) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: will be removed

File: sql/src/main/java/io/crate/metadata/cluster/DDLClusterStateHelpers.java [hunk 8]
```diff
@@ -90,4 +92,21 @@ static IndexTemplateMetaData templateMetaData(MetaData metaData, RelationName re
...
         return metaData.templates().get(templateName);
     }
+
+    static void renameIndices(MetaData metaData,
+                              MetaData.Builder mdBuilder,
+                              ClusterBlocks.Builder blocksBuilder,
+                              Index[] sourceIndices,
+                              String[] targetIndices) {
+        for (int i = 0; i < sourceIndices.length; i++) {
+            Index index = sourceIndices[i];
```
Comment (andreidan) [inline sql/src/main/java/io/crate/metadata/cluster/DDLClusterStateHelpers.java]: minor: I'd rename `index` to `source` or something similar as it's a bit hard to follow - especially as below at some point you see `mdBuilder.remove(index.getName());` and think .. which index is that? :)

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 9]
```diff
@@ -236,6 +284,211 @@ private Session session() {
...
+        String[] renameSourceNames = new String[2];
+        String[] renameTargetNames = new String[2];
+        renameSourceNames[0] = sourceIndexName;
+        renameTargetNames[0] = backupIndexName;
+        renameSourceNames[1] = targetIndexName;
+        renameTargetNames[1] = sourceIndexName;
+
+        String[] renameSourceNamesReverse = new String[2];
+        String[] renameTargetNamesReverse = new String[2];
+        renameTargetNamesReverse[0] = renameSourceNames[1];
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: maybe a personal preference, but this is very hard to follow, I'd rather have the names of the indexes be assigned (like above for `renameSourceNames` and `renameTargetNames`) and maybe extract this in a utility method that's unit tested

Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: I agree, it's one of the things that I do need to revise if we decide to keep this approach of renaming instead of exchange (in previous commit). What I don't like on this rename is that the operation is not symmetrical by nature so you need to do this processing in order to achieve the desired result. I was also considering using a different data structure to prepare the 'inversion' but I want to discuss first if it's worth going that way

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 6 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: removed

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 10]
```diff
@@ -236,6 +271,138 @@ private Session session() {
...
+        if (!blocks_write) {
+            throw new IllegalStateException("Table/Partition needs to be at a read-only state." +
+                                               " Use 'ALTER table ... set (blocks.write=true)' and retry");
+        }
+    }
+
+    private DiscoveryNode getNodeForResize() {
+        assert ! clusterService.state().nodes().getDataNodes().isEmpty(): "No Data nodes available " +
+                                                                          "for index resize operation";
+        return clusterService.state().nodes().getDataNodes().valuesIt().next();
```
[the remark this answers is not in the document]
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: I created a separate task for this - I think we should also take into consideration existing node allocation restrictions as then there will be no need to check that explicitly.

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 11]
```diff
@@ -236,6 +284,211 @@ private Session session() {
...
+            if (response.isAcknowledged()) {
+                activeShardsObserver.waitForActiveShards(request.indices(),
+                    ActiveShardCount.DEFAULT,
+                    request.ackTimeout(), shardsAcked -> {
+                        if (!shardsAcked && logger.isInfoEnabled()) {
+                            logger.info("[{}] indexes reallocation is in progress, " +
+                                        "but the operation timed out while waiting for " +
+                                        "enough shards to be back on green state. Timeout={}. " +
+                                        "Consider increasing the timeout or adding nodes to the cluster.",
+                                request.indices(), request.timeout());
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: I think we should either rollback the operation and log a warning or do nothing by best-effort (thus the shrink operation may fail).

File: sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+        MetaData.Builder mdBuilder = null;
+        ClusterBlocks.Builder blocksBuilder = ClusterBlocks.builder()
+            .blocks(currentState.blocks());
+
+        Index[] sourceIndices = indexNameExpressionResolver.concreteIndices(currentState,
+            STRICT_INDICES_OPTIONS, sourceIndexName);
+        Index[] sourceIndicesOrdered = new Index[sourceIndexName.length];
+
+        // ensure the order in input is respected
+        for (int i = 0; i < sourceIndexName.length; i++) {
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java]: Afaik this can be an assertion as `concreteIndices` will retain the order if we do not pass in aliases (which we don't)

File: sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    public RenameIndexClusterStateExecutor(Settings settings,
+                                           IndexNameExpressionResolver indexNameExpressionResolver) {
+        logger = Loggers.getLogger(getClass(), settings);
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+    }
+
+    @Override
+    public ClusterState execute(ClusterState currentState, RenameIndexRequest request) throws Exception {
+        final String[] sourceIndexName = request.sourceIndexName();
+        final String[] targetIndexName = request.targetIndexName();
```
Comment (seut) [inline sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java]: please use plural form of var names (`sourceIndicesNames`, `targetIndicesNames`)

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 14]
```diff
@@ -97,11 +111,16 @@
...
 
+import static org.elasticsearch.cluster.metadata.IndexMetaData.INDEX_BLOCKS_WRITE_SETTING;
+import static org.elasticsearch.cluster.metadata.IndexMetaData.INDEX_ROUTING_REQUIRE_GROUP_PREFIX;
+import static org.elasticsearch.cluster.metadata.IndexMetaData.SETTING_NUMBER_OF_SHARDS;
 import static org.elasticsearch.common.settings.AbstractScopedSettings.ARCHIVED_SETTINGS_PREFIX;
 
 @Singleton
 public class AlterTableOperation {
 
+    private static final String SHRINK_PREFIX = ".shrinked";
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: I'd prefer having a `.` at the end of the prefix as a (more human) separator.

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 15]
```diff
@@ -236,6 +284,211 @@ private Session session() {
...
+        ));
+
+        // close index
+        actions.add(new ChainableAction<>(
+            () -> closeTable(sourceIndexName, targetIndexName),
+            () -> openTable(sourceIndexName, targetIndexName)
+        ));
+
+        // index rename
+        final String backupIndexName = ".backup" + sourceIndexName;
```
Comment (seut) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: also here, I'd prefer to use a `.` human separator at the prefix end, also prefix could be a static var..

File: sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+    public RenameIndexClusterStateExecutor(Settings settings,
+                                           IndexNameExpressionResolver indexNameExpressionResolver) {
+        logger = Loggers.getLogger(getClass(), settings);
+        this.indexNameExpressionResolver = indexNameExpressionResolver;
+    }
+
+    @Override
+    public ClusterState execute(ClusterState currentState, RenameIndexRequest request) throws Exception {
+        final String[] sourceIndexName = request.sourceIndexName();
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java]: Got it, it's for a batch rename. In that case I think we should make the API more explicit and pair the source and index names better together. E.g. `request.renameActions()` where `renameActions` is  `List<RenameIndexOp>` with `RenameIndexOp` having a `source` and `target`

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: @mfussenegger yes, it will fail later. However does't that mean that we have already performed the shard allocation which might be an expensive operation and that in addition we have to undo it? If this holds, I would say that it makes sense to bail out before any of this occurs if we know that it will fail on a later stage anyway.

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 14 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: Afaik this is save as an index name starting with `.` will never be recognized as a partition or table. See https://github.com/crate/crate/blob/master/sql/src/main/java/io/crate/metadata/IndexParts.java#L156-L178

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 9 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: ok, we've decided to keep the BulcRename Operation thus the API will change - this was an init basic implementation until a decision was made

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 17]
```diff
@@ -614,4 +887,10 @@ public void fail(@Nonnull Throwable t) {
         }
 
     }
+
+    static void ensureNotNull(Object value, String message) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: There is `Objects.requireNonNull` for this in the stdlib

File: sql/src/main/java/io/crate/execution/ddl/index/TransportRenameIndexNameAction.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Override
+    public ClusterStateTaskExecutor<RenameIndexRequest> clusterStateTaskExecutor(RenameIndexRequest request) {
+        return executor;
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(RenameIndexRequest request, ClusterState state) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_WRITE,
+            indexNameExpressionResolver.concreteIndexNames(state, STRICT_INDICES_OPTIONS,
```
[the remark this answers is not in the document]
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/index/TransportRenameIndexNameAction.java]: great catch

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 19]
```diff
@@ -236,6 +284,211 @@ private Session session() {
         return result;
     }
 
+    private CompletableFuture<Long> executeAlterTableChangeNumberOfShards(AlterTableAnalyzedStatement analysis) {
```
[the remark this answers is not in the document]
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: we also need an `Optional<PartitionName>`, I think it's better to leave it as it is.

File: sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java (hunk 12 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java]: From what I've seen in `concreteIndices`, `Index` instances are added in a `HashSet` which is later turned into an array. As far as I understand, these operations will not enforce other of the input, you can have no guarantee on the order of the output array.

In any case, it is no longer relevant, code is removed.

File: sql/src/main/java/io/crate/metadata/cluster/DDLClusterStateHelpers.java [hunk 20]
```diff
@@ -90,4 +92,21 @@ static IndexTemplateMetaData templateMetaData(MetaData metaData, RelationName re
         String templateName = PartitionName.templateName(relationName.schema(), relationName.name());
         return metaData.templates().get(templateName);
     }
+
+    static void renameIndices(MetaData metaData,
```
[the remark this answers is not in the document]
Comment (smakalias) [reply sql/src/main/java/io/crate/metadata/cluster/DDLClusterStateHelpers.java]: removed

**Comment by smakalias** [comment]
@seut @mfussenegger most comments are now addressed - apart from some that we need to discuss to agree for the direction we want to take.

File: sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java [hunk 21]
```diff
@@ -22,69 +22,54 @@
...
+                              MetaData.Builder mdBuilder,
+                              ClusterBlocks.Builder blocksBuilder,
+                              String[] sourceIndexNames,
+                              String[] targetIndexNames) {
+        for (int i = 0; i < sourceIndexNames.length; i++) {
+            String sourceIndex = sourceIndexNames[i];
+            IndexMetaData indexMetaData = metaData.index(sourceIndex);
+            IndexMetaData targetIndexMetadata = IndexMetaData.builder(indexMetaData)
+                .index(targetIndexNames[i]).build();
+            mdBuilder.remove(sourceIndex);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java]: I think by creating the new metadata with all indices present and then mutating it we still have an order dependency here.

E.g.

```
rename actions:
  x -> y
  y -> z

results in:

mdBuilder.remove(x)
mdBuilder.put(mdForY)
mdBuilder.remove(y)
mdBuilder.put(mdForZ)
```

File: sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java [hunk 22]
```diff
@@ -22,69 +22,54 @@
...
-        // The MetaData will always be overridden (and not merged!) when applying it on a cluster state builder.
-        // So we must re-build the state with the latest modifications before we pass this state to possible modifiers
-        // or return it.
-        // Otherwise they would operate on the old MetaData and would just ignore any modifications.
-        currentState = ClusterState.builder(currentState).metaData(mdBuilder).blocks(blocksBuilder).build();
-        return currentState;
+    static void renameIndices(MetaData metaData,
+                              MetaData.Builder mdBuilder,
+                              ClusterBlocks.Builder blocksBuilder,
+                              String[] sourceIndexNames,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java]: Could we use the new the new RenameAction structure here? Otherwise we should at least add some basic sanity checks - that the arrays have the same size.

File: sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java [hunk 23]
```diff
@@ -220,4 +218,21 @@ private static void validateAllIndicesAreClosed(ClusterState currentState,
             }
         }
     }
+
+    private static void renameIndices(MetaData metaData,
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java]: Is this the same as in `RenameIndexClusterStateExecutor` ?

File: sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        @Nonnull String targetIndexName() {
+            return targetIndexName;
+        }
+    }
+
+    private List<RenameIndexAction> renameIndexActions;
+
+    public BulkRenameIndexRequest(int initialNumberOfRenameRequests) {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java]: Maybe a bit of nitpicking, but it would be nice if we could design the API of
the request class to be immutable - which we can enforce once some ES
components switch from `Streamable` over to `Writable`, there is progress going
on to make that happen. 

That would mean a constructor that takes the list of rename actions - and the
removal of the add method.

File: sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+            validationException = addValidationError("Rename Index Action list must not be empty", null);
+        }
+        return validationException;
+    }
+
+    public List<RenameIndexAction> renameIndexActions() {
+        return renameIndexActions;
+    }
+
+    public String[] sourceIndices() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java]: I think we should remove these here and have the call-site execute this logic if it is required. Or better have them use the `RenameIndexAction` structure directly. If we can use the type system to enforce a pairing of source and target we should do it, and not introduce a loose contract that two independent arrays must fit together.

File: sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java (hunk 24 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java]: ok, I had that in place at first and then changed it for convenience 's sake. 
But I can change this back, was not aware of the switch to `Writable` - thanks

File: sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java (hunk 25 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/index/BulkRenameIndexRequest.java]: Also something I had in place and changed it to be more in line of the existing APIs that operate on Arrays. I will change it

File: sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java (hunk 23 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java]: it's slightly different, it uses an `String[]` for sourceIndex instead of `Index[]`. 
We do not really need `Index` as we already have the concrete indices and also got rid of the array re-ordering (see previous commits)

Comment (smakalias) [reply sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java]: I've now changed it to operate on a `List<RenameIndexAction>`

File: sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java (hunk 22 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/metadata/cluster/RenameIndexClusterStateExecutor.java]: done

**Comment by smakalias** [comment]
retest this please

**Comment by smakalias** [comment]
retest this please

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: decided to keep all checks that we know they can cause the shrink operation to fail

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java (hunk 11 above)
Comment (smakalias) [reply sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: decided to bail out

**Comment by smakalias** [comment]
@seut @mfussenegger I believe all comments are now addressed,
for everything else we discussed to follow-up with separate PRs

Could you please check?

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 26]
```diff
@@ -463,19 +463,21 @@ private static DiscoveryNode getNodeForResize(ClusterState state) {
...
                         shardsAcked -> {
-                            if (!shardsAcked && logger.isInfoEnabled()) {
-                                logger.info("[{}] indexes shrinked, but the operation timed out while waiting for " +
+                            if (!shardsAcked) {
+                                logger.warn("[{}] indexes shrinked, but the operation timed out while waiting for " +
                                             "enough shards to be started. Timeout={}. " +
-                                            "Consider increasing the timeout or adding nodes to the cluster.",
+                                            "Consider increasing the timeout or adding nodes to the cluster and retry the operation",
                                     request.getTargetIndexRequest().indices(), request.timeout());
+                                throw new RuntimeException("Shrinking Request failed");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: I think we should be more specific here - The shink worked but timed out whle waiting for active shards. We can tell that to the user. And since we propagate that information to the user we don't have to log it.

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 27]
```diff
@@ -494,7 +496,12 @@ private static DiscoveryNode getNodeForResize(ClusterState state) {
 
     private CompletableFuture<Long> renameIndicesBulk(final List<RenameIndexAction> renameIndexActions) {
         BulkRenameIndexRequest request = new BulkRenameIndexRequest(renameIndexActions);
-        FutureActionListener<AcknowledgedResponse, Long> listener = new FutureActionListener<>(r -> 0L);
+        FutureActionListener<AcknowledgedResponse, Long> listener = new FutureActionListener<>(response -> {
+            if (!response.isAcknowledged()) {
+                throw new RuntimeException("Rename Index request was not acknowledged");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: Please include some more details what this means and what the user can do about it.

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 28]
```diff
@@ -464,20 +464,17 @@ private static DiscoveryNode getNodeForResize(ClusterState state) {
...
                         shardsAcked -> {
                             if (!shardsAcked) {
-                                logger.warn("[{}] indexes shrinked, but the operation timed out while waiting for " +
-                                            "enough shards to be started. Timeout={}. " +
-                                            "Consider increasing the timeout or adding nodes to the cluster and retry the operation",
-                                    request.getTargetIndexRequest().indices(), request.timeout());
-                                throw new RuntimeException("Shrinking Request failed");
+                                throw new RuntimeException("Shrinking operation was successful but the operation " +
+                                                           "timed out while waiting for enough shards to be started. " +
+                                                           "Consider increasing the timeout and then retry the operation");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: This doesn't make sense. The user can't increase the timeout and there is no point in retrying the operation. The shards will become active without user intervention (unless there are disk space issues or too few nodes, in which case the user has to resolve these issues instead of retrying the operation)

File: sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java [hunk 29]
```diff
@@ -464,20 +464,17 @@ private static DiscoveryNode getNodeForResize(ClusterState state) {
...
                         },
                         e -> {
                             throw new IllegalStateException(e);
                         });
                 } else {
-                    logger.warn("[{}] Publishing new cluster state during Shrink operation has timed out. Timeout={}",
-                        request.getTargetIndexRequest().indices(), request.timeout());
-                    throw new RuntimeException("Shrinking Request was not acknowledged");
+                    throw new RuntimeException("Publishing new cluster state during Shrink operation has timed out." +
+                                               "Consider increasing the timeout and then retry the operation");
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java]: Same here

File: sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java [hunk 30]
```diff
@@ -235,4 +218,21 @@ private static void validateAllIndicesAreClosed(ClusterState currentState,
...
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/metadata/cluster/RenameTableClusterStateExecutor.java]: Can we revert the change to move this method? Seems to be unrelated
